Adding cluster size to track extra

fmazzasc opened the pull request:
Follows the discussion presented in https://indico.cern.ch/event/1332124/
Cluster sizes are added to AO2Ds and cluster map is removed. Few modifications have been done to the ITS tracking and ITSTPC matching to get the cl size information (pinging @mconcas , @mpuccio , @shahor02 ). We have also updated the AnalysisDataModel and created a versioning of the trackExtra tables. Few dyn columns have been changed, I will undraft the PR as soon as I checked that we get consistent results with the old dataformat (pinging @ddobrigk , @jgrosseo )

mpuccio committed:
Add ITS cluster size to the AO2D

fmazzasc committed:
Add TracksExtra_001

fmazzasc committed:
Update readerhelper

fmazzasc committed:
Add cluster sizes to AB

fmazzasc committed:
Remove its cluster map

alibuild committed:
Please consider the following formatting changes

shahor02 reviewed (dismissed):
cannot judge about the backward compatibility of the AOD track model changes, the rest looks ok.

fmazzasc committed:
Merge pull request #18 from alibuild/alibot-cleanup-12044

Please consider the following formatting changes to #12044

jgrosseo commented on Framework/Core/include/Framework/AnalysisDataModel.h:
```diff
@@ -240,10 +241,11 @@ DECLARE_SOA_COLUMN(TrackPhiEMCAL, trackPhiEmcal, float);
 DECLARE_SOA_COLUMN(TrackTimeRes, trackTimeRes, float);                                        //! Resolution of the track time in ns (see TrackFlags::TrackTimeResIsRange)
 DECLARE_SOA_EXPRESSION_COLUMN(DetectorMap, detectorMap, uint8_t,                              //! Detector map: see enum DetectorMapEnum
-                              ifnode(aod::track::itsClusterMap > (uint8_t)0, static_cast<uint8_t>(o2::aod::track::ITS), (uint8_t)0x0) |
+                              ifnode(aod::track::itsClusterSizes > (uint32_t)0, static_cast<uint8_t>(o2::aod::track::ITS), (uint8_t)0x0) |
```
As DetectorMap is part of the old and the new version, we also need an old and a new version of this column

jgrosseo commented on Framework/Core/include/Framework/AnalysisDataModel.h:
```diff
@@ -390,7 +420,7 @@ DECLARE_SOA_EXTENDED_TABLE(TracksCovIU, StoredTracksCovIU, "TRACKCOV_IU", //! Tr
                            aod::track::C1Pt21Pt2);
 
-DECLARE_SOA_TABLE_FULL(StoredTracksExtra, "TracksExtra", "AOD", "TRACKEXTRA", //! On disk version of TracksExtra
+DECLARE_SOA_TABLE_FULL(StoredTracksExtra_000, "TracksExtra", "AOD", "TRACKEXTRA", //! On disk version of TracksExtra
```
```suggestion
DECLARE_SOA_TABLE_FULL(StoredTracksExtra_000, "TracksExtra", "AOD", "TRACKEXTRA", //! On disk version of TracksExtra, version 0
```

jgrosseo commented on Framework/Core/include/Framework/AnalysisDataModel.h:
```diff
@@ -408,9 +438,33 @@ DECLARE_SOA_TABLE_FULL(StoredTracksExtra, "TracksExtra", "AOD", "TRACKEXTRA", //
                        track::TrackEtaEMCAL, track::TrackPhiEMCAL, track::TrackTime, track::TrackTimeRes);
 
-DECLARE_SOA_EXTENDED_TABLE(TracksExtra, StoredTracksExtra, "TRACKEXTRA", //! Additional track information (clusters, PID, etc.)
+DECLARE_SOA_TABLE_FULL_VERSIONED(StoredTracksExtra_001, "TracksExtra", "AOD", "TRACKEXTRA", 1, // Version 1 of TracksExtra
```
```suggestion
DECLARE_SOA_TABLE_FULL_VERSIONED(StoredTracksExtra_001, "TracksExtra", "AOD", "TRACKEXTRA", 1, // On disk version of TracksExtra, version 1
```

jgrosseo commented on Framework/Core/include/Framework/AnalysisDataModel.h:
```diff
@@ -408,9 +438,33 @@ DECLARE_SOA_TABLE_FULL(StoredTracksExtra, "TracksExtra", "AOD", "TRACKEXTRA", //
+                                 track::TPCFractionSharedCls<track::TPCNClsShared, track::TPCNClsFindable, track::TPCNClsFindableMinusFound>,
+                                 track::TrackEtaEMCAL, track::TrackPhiEMCAL, track::TrackTime, track::TrackTimeRes);
+
+DECLARE_SOA_EXTENDED_TABLE(TracksExtra_000, StoredTracksExtra_000, "TRACKEXTRA", //! Additional track information (clusters, PID, etc.)
+                           track::DetectorMap);
+
+DECLARE_SOA_EXTENDED_TABLE(TracksExtra_001, StoredTracksExtra_001, "TRACKEXTRA", //! Additional track information (clusters, PID, etc.)
```
Does this work with the same name "TRACKEXTRA" ?

jgrosseo replied:
I guess the second needs to have _001 but I may be wrong

aalkin replied:
No, the `description` field should stay the same. The version is added to the tree name if the table is written out.

jgrosseo reviewed (changes requested):
Thanks for this work!

For now, I have only reviewed the changes in AnalysisDataModel. See my minor comments.

However, we have to split this PR in two PRs. The first one introduces the changes in the AnalysisDataModel **only**, but keeps version 0 as default.
Then when it is merged, you make the PR with the converter in O2Physics
Once that is merged, you make a PR which contains the rest and the change to version 1.

fmazzasc committed:
Add table versioning

fmazzasc committed:
Fix comments

Co-authored-by: Jan Fiete <jgrosseo@cern.ch>

fmazzasc committed:
Fix comments (2)

Co-authored-by: Jan Fiete <jgrosseo@cern.ch>

aalkin commented on Framework/Core/src/AODReaderHelpers.cxx:
```diff
@@ -176,7 +176,9 @@ AlgorithmSpec AODReaderHelpers::aodSpawnerCallback(std::vector<InputSpec>& reque
         } else if (description == header::DataDescription{"TRACKEXTRA"}) {
-          outputs.adopt(Output{origin, description, version}, maker(o2::aod::TracksExtraExtensionMetadata{}));
+          outputs.adopt(Output{origin, description, version}, maker(o2::aod::TracksExtra_000ExtensionMetadata{}));
+        } else if (description == header::DataDescription{"TRACKEXTRA_001"}) {
```
This does not seem correct, the version is encoded in subspec, not in description. I see that this is also the case for the mc particles table, which is also incorrect, it just works by accident because the extensions are identical. I will fix the case with mc particles, but here you should do the following
```cpp
if (description == header::DataDescription{"TRACKEXTRA"}) {
    if (version == 0U) {
    } else if (version == 1U) {
    }
}
```

fmazzasc committed:
Set old TrackExtra as default

fmazzasc committed:
Fix data descriptor for trackextra

alibuild committed:
Please consider the following formatting changes

fmazzasc committed:
Merge pull request #19 from alibuild/alibot-cleanup-12044

Please consider the following formatting changes to #12044

fmazzasc commented:
Thanks to all of you for the comments and the help.
- @jgrosseo : I implemented all your comments and switched to 000 as default. As soon as the converter in O2Physics is merged I will move to  001 
- @aalkin : I changed the data decription as you suggested

I also checked the itsClusterMap with 000 and 001, and the dynamic evaluation works smoothly (see picture). Opening the PR.
![Screenshot from 2023-10-11 12-37-57](https://github.com/AliceO2Group/AliceO2/assets/43742195/af77aa9d-f582-4e24-96af-ce5a8a4ed5f5)

fmazzasc committed:
Fix test

mconcas commented:
ITS side small changes also look ok.

jgrosseo commented:
Don't you want to factor our the AnalysisDataModel changes in a separate PR and then merge this one once you swap to version 1 and can activate the filling in the AOD producer instead of the commenting out code?

alibuild commented:
Error while checking build/O2/fullCI for 6282035bb82369c2d0cdd05bfa51bf18e0d100d3 at 2023-10-14 02:13:
```
No log files found
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/12044/6282035bb82369c2d0cdd05bfa51bf18e0d100d3/build_O2_fullCI/pretty.html).

fmazzasc committed:
Clean commented code

fmazzasc commented:
> Don't you want to factor our the AnalysisDataModel changes in a separate PR and then merge this one once you swap to version 1 and can activate the filling in the AOD producer instead of the commenting out code?

I removed the commented code and opened in draft mode https://github.com/AliceO2Group/AliceO2/pull/12087

jgrosseo commented on Framework/Core/include/Framework/AnalysisDataModel.h:
```diff
@@ -280,6 +304,24 @@ DECLARE_SOA_DYNAMIC_COLUMN(ITSNClsInnerBarrel, itsNClsInnerBarrel, //! Number of
                              }
                              return itsNclsInnerBarrel;
                            });
+DECLARE_SOA_DYNAMIC_COLUMN(ITSNCls001, itsNCls, //! Number of ITS clusters
```
I am sorry that we seem not to converge. ;-)

I think now where we have the `v001` namespace, we should move this there and call it `ITSNCls` and not `ITSNCls001`. The same for `ITSClusterMapDyn`. What do you think?

fmazzasc replied:
I agree. Done!

jgrosseo commented on Framework/Core/include/Framework/AnalysisDataModel.h:
```diff
@@ -221,7 +221,8 @@ DECLARE_SOA_EXPRESSION_COLUMN(C1Pt21Pt2, c1Pt21Pt2, float, //!
 DECLARE_SOA_COLUMN(Flags, flags, uint32_t);                                                   //! Track flags. Run 2: see TrackFlagsRun2Enum | Run 3: see TrackFlags
-DECLARE_SOA_COLUMN(ITSClusterMap, itsClusterMap, uint8_t);                                    //! ITS cluster map, one bit per a layer, starting from the innermost
+DECLARE_SOA_COLUMN(ITSClusterSizes, itsClusterSizes, uint32_t);                               //! Clusters sizes, four bits per a layer, starting from the innermost
+DECLARE_SOA_COLUMN(ITSClusterMap, itsClusterMap, uint8_t);                                    //! Old cluster ITS cluster map, kept for retrocompatibility
```
```suggestion
DECLARE_SOA_COLUMN(ITSClusterMap, itsClusterMap, uint8_t);                                    //! Old cluster ITS cluster map, kept for version 0 compatibility
```

jgrosseo commented on Framework/Core/test/test_AnalysisTask.cxx:
```diff
@@ -164,7 +164,7 @@ TEST_CASE("AdaptorCompilation")
   REQUIRE(task2.inputs[1].binding == "TracksExtension");
   REQUIRE(task2.inputs[2].binding == "Tracks");
-  REQUIRE(task2.inputs[3].binding == "TracksExtraExtension");
+  REQUIRE(task2.inputs[3].binding == "TracksExtra_000Extension");
```
This has to be changed again when we move to version 1. Is this the best way to implement the test?

fmazzasc replied:
Simple `TracksExtraExtension` does not work neither with 000 nor 001. Also for Collisions the test follows the same syntax

aalkin replied:
Binding can be indeed extracted from the type, I will update the test outside this PR.

fmazzasc committed:
Fix comment (3)

Co-authored-by: Jan Fiete <jgrosseo@cern.ch>

fmazzasc committed:
Move dyn columns to versioning namespace

fmazzasc committed:
Merge branch 'dev' into cl_size_ao2d

jgrosseo commented on Framework/Core/include/Framework/AnalysisDataModel.h:
```diff
@@ -239,11 +240,53 @@ DECLARE_SOA_COLUMN(TrackEtaEMCAL, trackEtaEmcal, float);
+
+namespace v001
+{
+DECLARE_SOA_EXPRESSION_COLUMN(DetectorMap, detectorMap, uint8_t, //! Detector map version 1
```
```suggestion
DECLARE_SOA_EXPRESSION_COLUMN(DetectorMap, detectorMap, uint8_t, //! Detector map version 1, see enum DetectorMapEnum
```

jgrosseo committed:
Update Framework/Core/include/Framework/AnalysisDataModel.h

fmazzasc commented:
@shahor02 , @aalkin , could we merge this PR?

shahor02 commented on Detectors/ITSMFT/ITS/tracking/include/ITStracking/TimeFrame.h:
```diff
@@ -244,6 +245,7 @@ class TimeFrame
   std::vector<float> mMSangles;
   std::vector<float> mPhiCuts;
   std::vector<float> mPositionResolution;
+  std::vector<int> mClusterSize;
```
why this needs to be int?

fmazzasc replied:
moved to unsigned short

shahor02 replied:
you anyway round it to max 15, right? Could go even to `int8_t`

fmazzasc replied:
Done. I needed to add an `if` for the treatement of overflows

fmazzasc replied:
Can we merge it now?

shahor02 commented on Detectors/GlobalTracking/include/GlobalTracking/MatchTPCITS.h:
```diff
@@ -601,12 +601,14 @@ class MatchTPCITS
+  gsl::span<const o2::its::TrackITS> mITSTracksArray;      ///< input ITS tracks span
+  gsl::span<const int> mITSTrackClusIdx;                   ///< input ITS track cluster indices span
+  std::vector<ITSCluster> mITSClustersArray;               ///< ITS clusters created in loadInput
+  std::vector<unsigned int> mITSClusterSizes;              ///< ITS cluster sizes created in loadInput
```
why this needs to be int?

fmazzasc replied:
Moved to unsigned short

shahor02 commented on DataFormats/Detectors/ITSMFT/ITS/include/DataFormatsITS/TrackITS.h:
```diff
@@ -122,13 +122,36 @@ class TrackITS : public o2::track::TrackParCov
 
+  void setClusterSize(int l, int size)
+  {
+    if (l >= 8)
+      return;
+    if (size > 15)
+      size = 15;
```
strange that this passed the check, could you put brackets?

fmazzasc replied:
Added

fmazzasc committed:
Add braces + unsigned shorts

fmazzasc committed:
Move cluster sizes to uint8_t

ktf commented on DataFormats/Detectors/ITSMFT/common/include/DataFormatsITSMFT/TrkClusRef.h:
```diff
@@ -26,11 +26,34 @@ namespace itsmft
 struct TrkClusRef : public o2::dataformats::RangeRefComp<4> {
   using o2::dataformats::RangeRefComp<4>::RangeRefComp;
   uint16_t pattern = 0; ///< layers pattern
+  uint32_t clsizes = 0; ///< cluster sizes for each layer
```
Shouldn't we swap the order of `pattern` and `clsizes` to avoid wasting 16 bits in padding per ref?

fmazzasc replied:
done, thanks for spotting!

fmazzasc committed:
Swap pattern and clsizes order

shahor02 commented:
checks were passed before the trivial swap, merging
